### ISSUE TITLE
GM2 Patch Change Overrides

### ISSUE THREAD

**Issue by JimBayne** [sinshu/meltysynth#31]
Many thanks for this wonderful product.

I am currently using it to render midi files to audio in a variety of ways.

I am stuck however, trying to override GM patch changes in the midi sequences with GM2 patch changes. I don't seem to be able to get the overrides to work.

I am using a SoundFont that claims to be GM2 compatible from https://musical-artifacts.com/artifacts/1346,
and have the following code in place ...

`        ... other stuff ...

        public const int PATCH_CHANGE = 0xC0;
        public const int BANK_CHANGE = 0xB0;
        public const int BANK_0 = 0;
        public const int BANK_32 = 32;

        ... other stuff ...

                // Create the synthesizer.
                synthesizer = new Synthesizer(SoundFontFile, sampleRate);

                midiFile = new MidiFile(MidiFile);
                sequencer = new MidiFileSequencer(synthesizer);

                sequencer.OnSendMessage = (synthesizer, channel, command, data1, data2) =>
                {
                    if (command == PATCH_CHANGE && channel >= 0 && channel <= 15 && PatchOverRide[channel] != NO_OVERRIDE)
                    {  //  sequencer encountered a patch change for a channel that we want to override the patch for. Override the patch ...
                        synthesizer.ProcessMidiMessage(channel, BANK_CHANGE, BANK_0, PatchOverRide_Bank_0_MSB[channel]);   // issue the Bank 0 change
                        synthesizer.ProcessMidiMessage(channel, BANK_CHANGE, BANK_32, PatchOverRide_Bank_32_MSB[channel]); // issue the Bank 32 change
                        synthesizer.ProcessMidiMessage(channel, command, PatchOverRide[channel], data2);  //                  issue the revised patch change  
                    }
                    else
                    {  //  We're not interested in this message. Let the sequencer process it "as is".
                        synthesizer.ProcessMidiMessage(channel, command, data1, data2);
                    }
                };

                sequencer.Play(midiFile, false);

                // setup the output buffer.
                left = new float[(int)(sampleRate * midiFile.Length.TotalSeconds)];
                right = new float[(int)(sampleRate * midiFile.Length.TotalSeconds)];

                // Render the waveform.
                sequencer.Render(left, right);

        ... other stuff ...`

The PatchOverRide's are working, however the BANK_0, and BANK_32 bank changes have no effect. For instance

Patch  Bank32  Bank0  Name
063    121     001    Synth Brass 1
063    121     002    Synth Brass 3
063    121     003    Analog Brass 1
063    121     004    Jump Brass 1

all result in the same instrument being selected (063    121     001    Synth Brass 1) - i.e. no difference in the sound.

Do you have any suggestions as to how I can overcome this issue?

Regards
Jim Bayne

**Comment by sinshu** [comment]
Could you share an example MIDI file that does not work well?

**Comment by JimBayne** [comment]
Thanks for the quick response. I have attached a sample midi file.

Try listening to / rendering the first 10 bars with the patch for channel 6  set to

              patch#  bank32  bank0
               063    121     001     Synth Brass 1
   ...         063    121     002     Synth Brass 3
   ...         063    121     003     Analog Brass 1
   ...         063    121     004     Jump Brass 1

I find they all sound identical, but ...

I don't thing MeltySynth is the problem. I have tried using Coolsoft's Virtual Midi Synth with the
GM2 compatible SoundFount from https://musical-artifacts.com/artifacts/1346  and Virtual Midi Synth 
has the same problem. All patch variants for the same patch # from the different banks sound the same.

It's most likely that the SoundFont has the same audio samples for the different banks.

Do you know of a SoundFont that does have a full set of GM2 samples?
[SampeMidiFile.zip](https://github.com/sinshu/meltysynth/files/10168780/SampeMidiFile.zip)

**Comment by JimBayne** [comment]
Many thanks. Your response was extremely helpful, and solved the problem for me. I had the bank 32 LSB and bank 0 MSB values reversed when I interpreted the patch maps. Once I sorted that out, my MeltySynth patch overrides began to work. I couldn't really hear the difference with the Synth-Brass patches, but when I tested with the Sitar patches it became very evident that I was obtaining different voicings.
